docs/code/introduction/mode_code.md

```diff
@@ -0,0 +1,126 @@
+
+## Add some custom methods to do things
+
+You can look at the Mode base class (the link from GitHub from earlier) to see what the base Mode class does. However, we have created a few “convenience” methods that you can use. They are:
+
+* **mode_init:** Called once when MPF is starting up
+
+* **mode_start:** Called every time the mode starts, just after the mode_<name>_started event is posted.
+
+* **add_mode_event_handler:** This is the same as the main add_event_handler() method from the Event Manager, except since it’s mode-specific it will also automatically remove any event handlers that you registered when the mode stops. (If you want to register event handlers that are always watching for events even when the mode is not running, you can use the regular self.machine.mode.add_handler() method.
+
+You don’t have to use all of these if you don’t want to.
+
+Also, modes have additional convenience attributes you can use within your mode code:
+
+* **self.config:** A link to the config dictionary for the mode’s config file.
+
+* **self.priority:** The priority the mode is running at. (Don’t change this. Just read it.)
+
+
+* **self.delay:** An instance of the delay manager you can use to set delayed callbacks for this mode. Any active ones will be automatically removed when the mode ends.
+
+* **self.player:** A link to the current player object that’s automatically updated when the player changes. This will be None if the mode is running outside of a game.
+
+* **self.active:** A boolean (True/False) value you can query to see if the mode is running.
+
+## Example usage
+
+Here’s an example of some mode code in use. This example is just a bunch of random things, but again, since you’re writing code here, the sky’s the limit! Seriously you could do all your game logic in mode code and not use the MPF configs at all if you wanted to.
+
+
+```
+from mpf.core.mode import Mode
+
+
+class Base(Mode):
+
+    def mode_init(self):
+        print("My custom mode code is being initialized")
+
+    def mode_start(self, **kwargs):
+        # The mode_start method needs **kwargs because some events that
+        # start modes pass additional parameters
+
+        print("My custom mode code is starting")
+
+        # call a delay in 5 seconds
+        self.delay.add(5000, self.my_callback)
+
+        # what player are we?
+        print(self.player.number)
+
+        # what's the player's score?
+        print('Score: {}'.format(self.player.score))
+
+        self.add_mode_event_handler('player_score', self.player_score_change)
+
+        # turn LED "led01" red
+        self.machine.leds.led01.color('red')
+
+    def my_callback(self):
+        print("My delayed call was just called!")
+
+    def player_score_change(self, **kwargs):
```

Comment:
This would be a good chance to use event kwargs, to show that event handlers have access to all event variables.

```
def player_score_change(self, **kwargs):
  self.machine.log.info("Player score went up by %s, was %s and is now %s", kwargs['change'], kwargs['prev_value'], kwargs['value'])
```

Those values could be defined as explicit kwargs in the function signature (`def player_score_change(value, prev_value, change, **kwargs)` but I don't know if that's more clear or less.

Reply:
I added the kwargs usage as an option that people see how to use it. I believe this way it should be good.